### PR TITLE
Pipeline triggers

### PR DESCRIPTION
**Pull request by jdoliner**
This PR extends triggers a little bit so they can be included directly in pipeline specs rather than needing to create the branch separately.

### CODE REVIEW

**Comment by jdoliner** [comment]
One thing that's arguably missing from this PR is that deleting or updating the pipeline doesn't clean up the trigger branches that it creates. This is maybe something that should be added but I was a little unsure for this PR. For one it's a little bit tricky to do since you have to figure out which branches to delete. Furthermore there's no guarantee that other pipelines aren't using the branches so you have to deal with that case. The behavior is also fairly in keeping with our behavior when the pipeline takes an input branch that doesn't yet exist. The branch gets created and its lifetime isn't tied to the lifetime of the pipeline, it seems a little different because most of the time that's creating a `master` branch and you generally don't want that cleaned up. But in this case it's a branch specific to the trigger. Ultimately though that branch consumes very few resources and it only clutters up `list branch` which isn't a commonly run command, so I think it's reasonable to punt on this for now and see if users find it annoying enough to justify fixing it.

File: doc/docs/master/concepts/advanced-concepts/deferred_processing.md [hunk 1]
```diff
@@ -284,6 +284,40 @@ When more than one is specified, a branch repoint will be triggered when any of
...
+  "transform": {
+    "cmd": [ "python3", "/edges.py" ],
+    "image": "pachyderm/opencv"
+  }
+}
+```
+
+When you create this pipeline Pachyderm will also create a branch in the input
+repo that specifies the trigger and the pipeline will use that branch as its
+input. The name of the branch is auto generated of the form
```
Comment (Tryneus) [inline doc/docs/master/concepts/advanced-concepts/deferred_processing.md:317]:
```suggestion
input. The name of the branch is auto-generated with the form
```

File: doc/docs/master/concepts/advanced-concepts/deferred_processing.md [hunk 2]
```diff
@@ -284,6 +284,40 @@ When more than one is specified, a branch repoint will be triggered when any of
...
+`<pipeline-name>-trigger-n` you can manually update the heads of these branches
```
Comment (Tryneus) [inline doc/docs/master/concepts/advanced-concepts/deferred_processing.md:318]:
```suggestion
`<pipeline-name>-trigger-n`. You can manually update the heads of these branches
```

File: doc/docs/master/reference/pipeline_spec.md [hunk 3]
```diff
@@ -518,8 +518,15 @@ single repo.
     "branch": string,
     "glob": string,
     "lazy" bool,
-    "empty_files": bool
-    "s3": bool
+    "empty_files": bool,
+    "s3": bool,
+    "trigger" : {
```
Comment (Tryneus) [inline doc/docs/master/reference/pipeline_spec.md:523]:
```suggestion
    "trigger": {
```

File: doc/docs/master/reference/pipeline_spec.md [hunk 4]
```diff
@@ -518,8 +518,15 @@ single repo.
...
+        "branch": string,
+        "all": bool,
+        "cron_spec": string,
+        "size": int,
```
Comment (Tryneus) [inline doc/docs/master/reference/pipeline_spec.md:527]: This is a `string`, I believe.

**Review by Tryneus — APPROVED** [review]
LGTM with a few nits.  Also it might be nice to test the overwriting behavior when updating a pipeline to have more or fewer triggers in the same input repo.

File: doc/docs/master/reference/pipeline_spec.md (hunk 4 above)
Comment (jdoliner) [reply doc/docs/master/reference/pipeline_spec.md:527]: Oops right you are.

File: doc/docs/master/reference/pipeline_spec.md [hunk 5]
```diff
@@ -518,8 +518,15 @@ single repo.
...
     "lazy" bool,
-    "empty_files": bool
-    "s3": bool
+    "empty_files": bool,
+    "s3": bool,
+    "trigger": {
+        "branch": string,
+        "all": bool,
+        "cron_spec": string,
+        "size": int,
```
Comment (jdoliner) [inline doc/docs/master/reference/pipeline_spec.md:527]:
```suggestion
        "size": string,
```

File: doc/docs/master/concepts/advanced-concepts/deferred_processing.md [hunk 6]
```diff
@@ -284,6 +284,40 @@ When more than one is specified, a branch repoint will be triggered when any of
 the conditions is met. To guarantee that they all must be met, add
 --trigger-all.
 
+## Embed Triggers in Pipelines
+
+Triggers can also be specified in the pipeline spec and automatically created
+when the pipeline is created. For example, this is the edges pipeline from our
+our OpenCV demo modified to only trigger when there is a MB of new images:
```
Comment (pappasilenus) [inline doc/docs/master/concepts/advanced-concepts/deferred_processing.md:291]: a MB ->  1 Megabyte

so we're consistent with prior paragraphs.

File: doc/docs/master/concepts/advanced-concepts/deferred_processing.md [hunk 7]
```diff
@@ -284,6 +284,40 @@ When more than one is specified, a branch repoint will be triggered when any of
...
+    }
+  },
+  "transform": {
+    "cmd": [ "python3", "/edges.py" ],
+    "image": "pachyderm/opencv"
+  }
+}
+```
+
+When you create this pipeline Pachyderm will also create a branch in the input
```
Comment (pappasilenus) [inline doc/docs/master/concepts/advanced-concepts/deferred_processing.md:315]: separate the introductory phrase with a comma:

When you create this pipeline,  Pachyderm will also create a branch in the input

**Review by pappasilenus — CHANGES_REQUESTED** [review]
Just a couple of copy-editing nits, otherwise looks clear to me.

File: doc/docs/master/concepts/advanced-concepts/deferred_processing.md (hunk 7 above)
Comment (jdoliner) [reply doc/docs/master/concepts/advanced-concepts/deferred_processing.md:315]:
```suggestion
When you create this pipeline, Pachyderm will also create a branch in the input
```

File: doc/docs/master/concepts/advanced-concepts/deferred_processing.md (hunk 6 above)
Comment (jdoliner) [reply doc/docs/master/concepts/advanced-concepts/deferred_processing.md:291]:
```suggestion
our OpenCV demo modified to only trigger when there is a 1 Megabyte of new images:
```

**Review by pappasilenus — APPROVED** [review]
Looks good!

**Pull request merged**